File: python/src/flakegate/report.py
from __future__ import annotations

import json
from typing import Any

from flakegate.runner import RunResult


def to_dict(run: RunResult) -> dict[str, Any]:
    return {
        "passed": run.passed,
        "cases": [
            {
                "id": r.case_id,
                "repeat": len(r.raw_responses) + len(r.errors),
                "successful_calls": len(r.raw_responses),
                "failed_calls": len(r.errors),
                "consistency_score": round(r.score.consistency_score, 4),
                "threshold": r.threshold,
                "passed": r.score.passed,
                "distinct_answers": r.score.distinct_answers,
                "majority_answer": r.score.majority_answer,
                "distribution": r.score.distribution,
                "prompt_tokens": r.prompt_tokens,
                "completion_tokens": r.completion_tokens,
            }
            for r in run.case_results
        ],
    }


def to_json(run: RunResult) -> str:
    return json.dumps(to_dict(run), indent=2)


def to_markdown(run: RunResult) -> str:
    status = "PASSED" if run.passed else "FAILED"
    lines = [
        f"## Flakegate report — {status}",
        "",
        "| Case | k | Distinct answers | Consistency | Threshold | Result |",
        "|---|---|---|---|---|---|",
    ]
    for r in run.case_results:
        k = len(r.raw_responses) + len(r.errors)
        mark = "PASS" if r.score.passed else "FAIL"
        lines.append(
            f"| `{r.case_id}` | {k} | {r.score.distinct_answers} | "
            f"{r.score.consistency_score:.0%} | {r.threshold:.0%} | {mark} |"
        )
        if r.errors:
            lines.append(f"|  |  |  |  |  | {len(r.errors)} call(s) errored |")
    return "\n".join(lines) + "\n"


def render(run: RunResult, fmt: str) -> str:
    if fmt == "json":
        return to_json(run)
    if fmt == "md":
        return to_markdown(run)
    raise ValueError(f"unknown report format: {fmt}")

File: node/src/report.ts
import type { CaseResult, RunResult } from "./runner.js";

/**
 * JSON report keys are snake_case (matching the Python CLI's output and the
 * config file's own convention) so tooling can consume a report from either
 * language's CLI identically.
 */
export interface CaseReport {
  id: string;
  repeat: number;
  successful_calls: number;
  failed_calls: number;
  consistency_score: number;
  threshold: number;
  passed: boolean;
  distinct_answers: number;
  majority_answer: string;
  distribution: Record<string, number>;
  prompt_tokens: number;
  completion_tokens: number;
}

export interface RunReport {
  passed: boolean;
  cases: CaseReport[];
}

function toCaseReport(r: CaseResult): CaseReport {
  return {
    id: r.caseId,
    repeat: r.rawResponses.length + r.errors.length,
    successful_calls: r.rawResponses.length,
    failed_calls: r.errors.length,
    consistency_score: Math.round(r.score.consistencyScore * 10000) / 10000,
    threshold: r.threshold,
    passed: r.score.passed,
    distinct_answers: r.score.distinctAnswers,
    majority_answer: r.score.majorityAnswer,
    distribution: r.score.distribution,
    prompt_tokens: r.promptTokens,
    completion_tokens: r.completionTokens,
  };
}

export function toReport(run: RunResult): RunReport {
  return { passed: run.passed, cases: run.caseResults.map(toCaseReport) };
}

export function toJson(run: RunResult): string {
  return JSON.stringify(toReport(run), null, 2);
}

export function toMarkdown(run: RunResult): string {
  const status = run.passed ? "PASSED" : "FAILED";
  const lines = [
    `## Flakegate report — ${status}`,
    "",
    "| Case | k | Distinct answers | Consistency | Threshold | Result |",
    "|---|---|---|---|---|---|",
  ];
  for (const r of run.caseResults) {
    const k = r.rawResponses.length + r.errors.length;
    const mark = r.score.passed ? "PASS" : "FAIL";
    lines.push(
      `| \`${r.caseId}\` | ${k} | ${r.score.distinctAnswers} | ` +
        `${(r.score.consistencyScore * 100).toFixed(0)}% | ${(r.threshold * 100).toFixed(0)}% | ${mark} |`,
    );
    if (r.errors.length > 0) {
      lines.push(`|  |  |  |  |  | ${r.errors.length} call(s) errored |`);
    }
  }
  return `${lines.join("\n")}\n`;
}

export function render(run: RunResult, fmt: "json" | "md"): string {
  if (fmt === "json") return toJson(run);
  if (fmt === "md") return toMarkdown(run);
  throw new Error(`unknown report format: ${fmt}`);
}
